cmd/cronosd/cmd/fix-unlucky-tx.go

```diff
@@ -0,0 +1,223 @@
+			ctx := server.GetServerContextFromCmd(cmd)
+			clientCtx := client.GetClientContextFromCmd(cmd)
+
+			minBlockHeight, err := cmd.Flags().GetInt(FlagMinBlockHeight)
+			if err != nil {
+				return err
+			}
+
+			chainID, err := cmd.Flags().GetString(flags.FlagChainID)
+			if err != nil {
+				return err
+			}
+
+			var blocksFile io.Reader
+			if args[0] == "-" {
+				blocksFile = os.Stdin
+			} else {
+				fp, err := os.Open(args[0])
+				if err != nil {
+					return err
+				}
+				defer fp.Close()
+				blocksFile = fp
+			}
+
+			tmDB, err := openTMDB(ctx.Config, chainID)
+			if err != nil {
+				return err
+			}
+
+			scanner := bufio.NewScanner(blocksFile)
+			for scanner.Scan() {
+				blockNumber, err := strconv.ParseInt(scanner.Text(), 10, 64)
+				if err != nil {
+					return err
+				}
+
+				if blockNumber < int64(minBlockHeight) {
+					return fmt.Errorf("block number is generated before v0.7.0 upgrade: %d", blockNumber)
+				}
+
+				// load results
+				blockResults, err := tmDB.stateStore.LoadABCIResponses(blockNumber)
+				if err != nil {
+					return err
+				}
+
+				// find unlucky tx
+				var txIndex int64
+				for i, txResult := range blockResults.DeliverTxs {
+					if TxExceedsBlockGasLimit(txResult) {
+						if len(txResult.Events) > 0 && txResult.Events[len(txResult.Events)-1].Type == evmtypes.TypeMsgEthereumTx {
+							// already patched
+							break
+						}
+
+						// load raw tx
+						blk := tmDB.blockStore.LoadBlock(blockNumber)
+						if blk == nil {
+							return fmt.Errorf("block not found: %d", blockNumber)
+						}
+
+						tx, err := clientCtx.TxConfig.TxDecoder()(blk.Txs[i])
+						if err != nil {
+							return err
+						}
+
+						txIndex++
+						for msgIndex, msg := range tx.GetMsgs() {
+							ethTxIndex := txIndex + int64(msgIndex)
+							ethTx, ok := msg.(*evmtypes.MsgEthereumTx)
+							if !ok {
+								continue
+							}
+							evt := abci.Event{
+								Type: evmtypes.TypeMsgEthereumTx,
+								Attributes: []abci.EventAttribute{
+									{Key: []byte(evmtypes.AttributeKeyEthereumTxHash), Value: []byte(ethTx.Hash), Index: true},
+									{Key: []byte(evmtypes.AttributeKeyTxIndex), Value: []byte(strconv.FormatInt(ethTxIndex, 10)), Index: true},
+								},
+							}
+							txResult.Events = append(txResult.Events, evt)
+						}
+						if err := tmDB.stateStore.SaveABCIResponses(blockNumber, blockResults); err != nil {
+							return err
+						}
+						if err := tmDB.txIndexer.Index(&abci.TxResult{
+							Height: blockNumber,
+							Index:  uint32(i),
+							Tx:     blk.Txs[i],
+							Result: *txResult,
+						}); err != nil {
+							return err
+						}
+						for _, msg := range tx.GetMsgs() {
+							fmt.Println("patched", blockNumber, msg.(*evmtypes.MsgEthereumTx).Hash)
+						}
+						break
+					} else if txResult.Code == 0 {
+						// find the correct tx index
+						for _, evt := range txResult.Events {
+							if evt.Type == evmtypes.TypeMsgEthereumTx {
```

Comment:
why do you need to look for the tx index? is it to double check that that it can be found in the event or because it is not ordered?

Reply:
humm it can be different from i?

Reply:
yes, `i` is just the cosmos tx index, but we need the eth tx index, there are many cases makes them different, failed cosmos tx, non-eth cosmos tx, or batch tx.